android/quest/src/main/java/org/smartregister/fhircore/quest/ui/questionnaire/QuestionnaireActivity.kt

```diff
@@ -107,6 +124,94 @@ class QuestionnaireActivity : BaseMultiLanguageActivity() {
     )
   }
 
+  fun setupLocationServices() {
+    if (viewModel.applicationConfiguration.logQuestionnaireLocation) {
+      fusedLocationClient = LocationServices.getFusedLocationProviderClient(this)
+
+      if (!LocationUtils.isLocationEnabled(this)) {
+        openLocationServicesSettings()
+      }
+
+      if (!hasLocationPermissions()) {
+        launchLocationPermissionsDialog()
+      }
+
+      if (LocationUtils.isLocationEnabled(this) && hasLocationPermissions()) {
+        fetchLocation(true)
+      }
+    }
+  }
+
+  fun hasLocationPermissions(): Boolean {
+    return PermissionUtils.checkPermissions(
+      this,
+      listOf(
+        Manifest.permission.ACCESS_COARSE_LOCATION,
+        Manifest.permission.ACCESS_FINE_LOCATION,
+      ),
+    )
+  }
+
+  fun openLocationServicesSettings() {
+    activityResultLauncher =
+      PermissionUtils.getStartActivityForResultLauncher(this) { resultCode, _ ->
+        if (resultCode == RESULT_OK || hasLocationPermissions()) {
+          fetchLocation()
+        }
+      }
+
+    val intent = Intent(Settings.ACTION_LOCATION_SOURCE_SETTINGS)
+    showLocationSettingsDialog(intent)
+  }
+
+  private fun showLocationSettingsDialog(intent: Intent) {
+    AlertDialog.Builder(this)
+      .setMessage(getString(R.string.location_services_disabled))
+      .setCancelable(true)
+      .setPositiveButton(getString(R.string.yes)) { _, _ -> activityResultLauncher.launch(intent) }
+      .setNegativeButton(getString(R.string.no)) { dialog, _ -> dialog.cancel() }
+      .show()
+  }
+
+  fun launchLocationPermissionsDialog() {
+    locationPermissionLauncher =
+      PermissionUtils.getLocationPermissionLauncher(
+        this,
+        onFineLocationPermissionGranted = { fetchLocation(true) },
+        onCoarseLocationPermissionGranted = { fetchLocation(false) },
+        onLocationPermissionDenied = {
+          Toast.makeText(
+              this,
+              getString(R.string.location_permissions_denied),
+              Toast.LENGTH_SHORT,
+            )
+            .show()
+          Timber.e("Location permissions denied")
+        },
+      )
+
+    locationPermissionLauncher.launch(
+      arrayOf(
+        Manifest.permission.ACCESS_FINE_LOCATION,
+        Manifest.permission.ACCESS_COARSE_LOCATION,
+      ),
+    )
+  }
+
+  fun fetchLocation(highAccuracy: Boolean = true) {
+    lifecycleScope.launch {
+      try {
+        if (highAccuracy) {
+          currLocation = LocationUtils.getAccurateLocation(fusedLocationClient)
+        } else {
+          currLocation = LocationUtils.getApproximateLocation(fusedLocationClient)
+        }
+      } catch (e: Exception) {
+        Timber.e(e, "Failed to get GPS location")
```

Comment:
If this method only fetches locations for the questionnaire could we augment the Timber log and add the QR Id on for the QR that failed to get the GPS attached